Auth: 비밀번호 변경

wch-os opened the pull request:
## Issues
- Resolves #79 
- Resolves #413 
- Resolves #414
 
## Description

- 비밀번호 변경 기능 구현

- [POST] `/password/verify`
  - 로그인 유저가 비밀번호 변경 전, 본인 확인을 위해 비밀번호를 재인증

- [POST] `/password/change`
  - 로그인 및 재인증한 유저에 한해 비밀번호 변경 진행

<br />

## Review Points

- 자유로운 리뷰 부탁드립니다.

<br />

## How Has This Been Tested?

- 포스트맨 API 테스트로 비밀번호 재인증과 비밀번호 변경 기능 동작을 확인했습니다.

<br />

## Additional Notes

- 기능을 명확하게 구분하기 위해 사용자가 비로그인 상태에서 비밀번호를 잊었을 때 사용하는 기능인 **비밀번호 찾기(재설정)**, 사용자가 로그인 상태에서 마이페이지를 통해 진행되는 **비밀번호 변경** 기능을 별도의 이슈로 분리했습니다.
- 사용자의 안전한 비밀번호 설정을 위한, **비밀번호 히스토리 관리 기능**은 이후에 고도화를 진행하려 합니다.

wch-os committed:
fix(auth): make sendPasswordResetEmail endpoint return void

wch-os committed:
feat(auth): implement password verification and change usecases in application layer

usecase 정의
- verifyPassword: 비밀번호 변경 전, 기존 비밀번호 검증
- changePassword: 비밀번호 변경

port 추가 및 정의
- 인가 객체의 userId로부터 userEntity를 조회하기 위한 findById 추가
  (조회한 userEntity의 비밀번호 변경)

wch-os committed:
feat(auth): implement findById method in rdb layer

wch-os committed:
feat(auth): define password verification and change APIs in web layer

비밀번호 변경 전 재인증
- POST: /password/verify

비밀번호 변경
- POST: /password/change

wch-os committed:
fix(auth): add slash to password reset endpoint paths

merge-simpson commented on services/auth/application/src/main/java/nettee/auth/port/AuthQueryRepositoryPort.java:
```diff
@@ -4,6 +4,7 @@
 import nettee.auth.domain.User;
 
 public interface AuthQueryRepositoryPort {
+    Optional<User> findById(String userId);
```
**💊 Command 포트를 사용하는 건 어떨까요?**

이번에도 마찬가지로, 민감한 작업을 하기 때문에 조회보다는 명령 쪽(원본 핸들링 쪽)의 조회 기능을 사용하는 게 좋아 보입니다!

wch-os replied:
좋은 제안 감사합니다!
이전에 코멘트 남겨주신 내용도 반영해서 Command 포트에서의 작업으로 수정했습니다!

merge-simpson commented on services/auth/api/exception/src/main/java/nettee/auth/exception/AuthErrorCode.java:
```diff
@@ -27,6 +27,7 @@ public enum AuthErrorCode implements ErrorCode {
     AUTH_OTP_DESERIALIZE_FAILED("OTP 역직렬화에 실패하였습니다.", HttpStatus.INTERNAL_SERVER_ERROR),
     AUTH_PASSWORD_RESET_INVALID("유효하지 않은 비밀번호 변경 요청입니다.", HttpStatus.UNAUTHORIZED),
     AUTH_REFRESH_TOKEN_NOT_FOUND("리프레시 토큰을 찾을 수 없습니다.", HttpStatus.UNAUTHORIZED),
+    AUTH_PASSWORD_NOT_MATCH("비밀번호가 일치하지 않습니다.", HttpStatus.UNAUTHORIZED),
```
별건 아니지만 `AUTH_USER_PASSWORD_MISMATCHED` 등 이름도 사용할 수 있어 보입니다!

merge-simpson commented on services/auth/driving/web-mvc/src/main/java/nettee/auth/web/AuthCommandApi.java:
```diff
@@ -129,17 +131,17 @@ public ResponseEntity<String> verifyEmail(@RequestBody EmailVerifyRequest reques
 
-    @PostMapping("email/password/reset/send")
+    @PostMapping("/email/password/reset/send")
     @Operation(
             summary = "이메일 비밀번호 재설정 링크 전송",
             description = "사용자의 이메일로 비밀번호를 변경할 수 있는 링크를 발송합니다."
     )
     @ResponseStatus(HttpStatus.NO_CONTENT)
-    public ResponseEntity<String> sendPasswordResetEmail(@RequestBody PasswordForgotRequest request) {
+    public void sendPasswordResetEmail(@RequestBody PasswordForgotRequest request) {
```
두 가지 선택지가 있는 것 같습니다:

1. `@ResponseStatus(HttpStatus.NO_CONENT)`로 충분할 때  
    ```java
    // ResponseEntity와 중복되게 사용하지 않아도 됩니다.
    public void ...
    ```
2. ResponseEntity<Void> 사용 시  
    - @ResponseStatus(HttpStatus.NO_CONENT) 제거
    - .noContent() 등으로 반환

---

만약 코파일럿의 제안대로 기존 작업 방식처럼 `ResponseEntity`를 사용한다면
이처럼 `@ResponseStatus(HttpStatus.NO_CONENT)`를 제거할 수 있어 보입니다.

```suggestion
    @PostMapping("/email/password/reset/send")
    @Operation(
            summary = "이메일 비밀번호 재설정 링크 전송",
            description = "사용자의 이메일로 비밀번호를 변경할 수 있는 링크를 발송합니다."
    )
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public ResponseEntity<Void> sendPasswordResetEmail(@RequestBody PasswordForgotRequest request) {
```

merge-simpson reviewed (approved):
빠른 작업 감사합니다! 👍 
작업 내용이 잘 정리되어 있어서 쉽게 확인했습니다!

꼼꼼하게 잘 작성된 것 같아요.
미리 Approve 해 두겠습니다!

한두 가지 제안이 있는데 다음에 논의하면 좋을 것 같고, 이번에 반영하지 않아도 괜찮을 것 같습니다!

wch-os committed:
refactor(auth): renames AuthErrorCode for clarity

AUTH_PASSWORD_NOT_MATCH → AUTH_PASSWORD_MISMATCHED

wch-os committed:
refactor(auth): reloacate user queries to `CommandPort` for data consistency

비밀번호 재설정과 같은 작업에서 추후 Read DB와 Command DB 분리 시, user 정보의 정합성을 지키기 위해 Query Port에서 Command Port로 수정함

wch-os committed:
Merge remote-tracking branch 'origin/develop' into feature/auth-password-change
